config/m4/ucg.m4

```diff
@@ -0,0 +1,34 @@
+#
+# Copyright (C) Huawei Technologies Co., Ltd. 2019.  ALL RIGHTS RESERVED.
+# Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+# See file LICENSE for terms.
+#
+
+#
+# Enable UCG - Group collective operations component
+#
+
+ucg_modules=""
+
+m4_sinclude([src/ucg/configure.m4]) # m4_sinclude() silently ignores errors
+
+#
+# TODO this section can be moved inside src/ucg/configure.m4, which will be
+# included only if exists. In this case m4_esyscmd() and m4_if() wrapper can
+# also be removed.
+#
+AC_ARG_ENABLE([ucg],
+              AS_HELP_STRING([--enable-ucg],
+                             [Enable the group collective operations component, default: NO]),
+              [],
+              [enable_ucg=no])
+AS_IF([test "x$enable_ucg" != xno],
+      [ucg_modules=":builtin"
+       AC_DEFINE([ENABLE_UCG], [1],
+                 [Enable Groups and collective operations support (UCG)])
+       AC_MSG_NOTICE([Building with Groups and collective operations support (UCG)])
+      ])
+AS_IF([test -f ${ac_confdir}/src/ucg/Makefile.am],
+      [AC_SUBST([UCG_SUBDIR], [src/ucg])])
+
+AM_CONDITIONAL([HAVE_UCG], [test "x$enable_ucg" != xno])
```

Comment:
can we move this inside src/ucg/configure.m4 as well?
so that ucx will use m4_sinclude() for all UCG m4 part and ignore if doesn't exist

Reply:
@yosefe I don't like the idea of putting this inside UCG. One reason is that the roadmap includes dependence of ucg on other ucx components. Another is to allow ucg submodule to be easily configurable from both within UCX and the new UCF collective library.

Reply:
@alex--m AFAIK the ucx+ucg bundle would behave the same way if the AC_ARG_ENABLE+AS_IF constructs are in src/ucg/configure.m4 since ucg.m4 includes that file verbatim, what am i missing?

Reply:
@yosefe i'm not saying it will not work (it will), but what you are proposing is to move ucx-specific way to include ucg into ucg itself. Aside from being wrong from a code-organisation point of view - it may also interfere with my plan to integrate ucg in the new collective library.

Reply:
@alex--m IMHO the following code block (except maybe `ucg_modules=":builtin"`) is pretty generic and can make UCG fit in any library:
```
AC_ARG_ENABLE([ucg],
              AS_HELP_STRING([--enable-ucg],
                             [Enable the group collective operations component, default: NO]),
              [],
              [enable_ucg=no])
AS_IF([test "x$enable_ucg" != xno],
      [ucg_modules=":builtin"
       AC_DEFINE([ENABLE_UCG], [1],
                 [Enable Groups and collective operations support (UCG)])
       AC_MSG_NOTICE([Building with Groups and collective operations support (UCG)])
      ])
```

Reply:
As I discussed with yossi offline - I don't see what is the technical justification for this change (_taking into account that the current state of things is working_). On the other hand, this breaks the way I plan to integrate UCG into UCC. Essentially this creates an unnecessary constraint where the configure code which integrates UCG inside UCX must also be used for the UCC integration (by making the configure code reside in UCG).

Reply:
how to include uc*g* functionality into uc*c* project is out of scope of this PR and should be discussed in scope of UCC WG.
UCX wise - ucg.m4 file should be part of UCG project and included from configure.ac.

Reply:
@mike-dubman I don't see justification here - the claim ucg.m4 should reside in UCG is unsupported. On the contrary - UCX houses `rocm.m4` and `cuda.m4` rather then placing them in ROCm and CUDA respectively. Surely you're not suggesting these m4 files should (by this logic) be moved into their respective libraries?

Configure/m4 scripts for integration reside in the including project, not the included project (true for UCX and open-source projects in general). There should be a strong technical argument to do otherwise.

Reply:
* ucx official api relies on rocm, cuda and others hence it is hosted in ucx (as well as glibc, etc..)
* ucg adds and exports unofficial and experimental `collectives` api as temporary measure, while the official `collectives` api is discussed by WG and will be hosted in separate UCC project.
* hosting and offering ucg.m4 file makes ucx officialy and uncoditionally exporting collectives APIs which is goal of UCC, not UCX.

Reply:
@mike-dubman 
I don't see how this is a technical justification either. I agree with the first two items, but the third is a point of view I don't share: 

1. "official" is what the community decides as official. Are experimental headers official? is the UCT API official? both are present in UCX. In my point of view, the presence of `config/m4/ucg.m4` doesn't make it official. The UCG API will not be released in UCX's "official" release (see next item), so i'd say it's as far from official as one may get. Community consensus will determine what's official, and UCG is not there.

2. I'm sorry, but "unconditionally exporting collectives API" is just not true: Mellanox made an unprecedented request to require a flag to `autogen.sh` (never seen this happen in any project), in addition to a `configure` flag, in order to build UCG and export it's API (I agreed, because I thought it would speed up the merge process and avoid this kind of arguments). If either flag is not explicitly specified by the user - UCX will not build nor export the collectives API. Looks pretty conditional to me :)

Are there any technical justifications to favor this proposed amendment over the existing (working) patch?

Reply:
autogen flags are pretty standard https://github.com/open-mpi/ompi/blob/master/autogen.pl#L1138
anyway, spoke with @yosefe on this and agree that separate file is no concern. +0.5 from me :)